xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3966,6 +3966,50 @@ static std::optional<GlobalDeviceId> DeviceConstraint(
   return std::nullopt;
 }
 
+absl::Status IrEmitterUnnested::EmitCopyStartThunk(
+    const HloCopyStartInstruction* instr) {
+
+  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice dst_buffer,
+                      GetAllocationSliceForHlo(instr, {}));
+
+  const HloInstruction* src = instr->operand(0);
+  const Shape& input_shape = src->shape();
+  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice src_buffer,
+                      GetAllocationSliceForHlo(src, {}));
+  Shape shape = instr->shape();
+  CHECK(shape.IsTuple());
+
+  MemoryCopyType memory_copy_type = MemoryCopyType::kDeviceToDevice;
```

Comment:
Maybe instead of the pattern
```
Type t = defaultValue;
if (a) {
  t = 1;
}
if (b) {
  t = 2;
}
TF_RET_CHECK(t != defaultValue);
```
You could not need to rely on a default value which is unused by doing the following:
```
Type t;
if (a) {
  t = 1;
} else if (b) {
  t = 2;
} else {
  return Internal("Good error message");
}
```
That is, assuming the two are mutually exclusive.

Reply:
In fact, I think they're not mutually exclusive, you could have an async copy from host to host. 
Your code will set `memory_copy_type` to `kHostToDevice`.

Reply:
In this context, the two are mutually exclusive because the copy_start can generate host-to-device or device-to-host, no device-to-device. Thanks for showing me the pattern and I'm going to change correspondingly.